File: kernel/api/vector.go
// SiYuan - Refactor your thinking
// Copyright (c) 2020-present, b3log.org
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package api

import (
	"encoding/json"
	"net/http"
	"path/filepath"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/siyuan-note/siyuan/kernel/util"
	"github.com/siyuan-note/siyuan/kernel/vectordb"
)

// isEmbeddingReservedCollection 检查是否为 embedding 专用集合
// 这些集合只能通过 embedding 包的接口修改，通用 vector API 禁止直接操作
func isEmbeddingReservedCollection(name string) bool {
	return strings.HasPrefix(name, "blocks_embedding_") ||
		strings.HasPrefix(name, "assets_embedding_")
}

// ensureVectorDB 确保向量数据库已初始化
func ensureVectorDB() {
	if vectordb.GlobalDB != nil {
		return
	}
	// 设置 SSD 检测结果
	vectordb.IsSSD = util.IsWorkspaceOnSSD()
	// 初始化向量数据库到 data/storage/vectordb
	dbPath := filepath.Join(util.DataDir, "storage", "vectordb")
	vectordb.InitGlobalDB(dbPath)
}

func vectorBuildCollection(c *gin.Context) {
	ensureVectorDB()

	ret := map[string]interface{}{"code": 0}
	defer c.JSON(http.StatusOK, ret)

	body := map[string]interface{}{}
	if err := c.BindJSON(&body); err != nil {
		ret["code"] = -1
		ret["msg"] = "参数解析失败"
		return
	}

	collectionName, _ := body["collection_name"].(string)
	dimension := 0
	if d, ok := body["dimension"].(float64); ok {
		dimension = int(d)
	}

	if collectionName == "" {
		ret["code"] = -1
		ret["msg"] = "collection_name 不能为空"
		return
	}

	// 保护 embedding 专用集合名称
	if isEmbeddingReservedCollection(collectionName) {
		ret["code"] = 403
		ret["msg"] = "Embedding 集合名称受保护，请使用 /api/embedding/* 接口"
		return
	}

	if dimension <= 0 {
		ret["code"] = -1
		ret["msg"] = "dimension 必须大于 0"
		return
	}

	// 创建数据集
	col, err := vectordb.GlobalDB.CreateCollection(collectionName, dimension)
	if err != nil {
		ret["code"] = -1
		ret["msg"] = "创建数据集失败: " + err.Error()
		return
	}

	// 保存
	if err := vectordb.SaveCollection(col, vectordb.GlobalDB.Path); err != nil {
		ret["code"] = -1
		ret["msg"] = "保存数据集失败: " + err.Error()
		return
	}

	ret["data"] = map[string]interface{}{
		"collection_name": collectionName,
		"dimension":       dimension,
	}
}

func vectorAdd(c *gin.Context) {
	ensureVectorDB()

	ret := map[string]interface{}{"code": 0}
	defer c.JSON(http.StatusOK, ret)

	body := map[string]interface{}{}
	if err := c.BindJSON(&body); err != nil {
		ret["code"] = -1
		ret["msg"] = "参数解析失败"
		return
	}

	collectionName, _ := body["collection_name"].(string)
	pointsRaw, _ := body["points"].([]interface{})

	// 保护 embedding 专用集合
	if isEmbeddingReservedCollection(collectionName) {
		ret["code"] = 403
		ret["msg"] = "Embedding 集合受保护，请使用 /api/embedding/* 接口"
		return
	}

	col := vectordb.GlobalDB.GetCollection(collectionName)
	if col == nil {
		ret["code"] = -1
		ret["msg"] = "数据集不存在"
		return
	}

	addedCount := 0
	for _, pRaw := range pointsRaw {
		p, ok := pRaw.(map[string]interface{})
		if !ok {
			continue
		}
		id, _ := p["id"].(string)
		vectorRaw, _ := p["vector"].([]interface{})

		// 转换向量
		vec := make([]float32, len(vectorRaw))
		for i, val := range vectorRaw {
			if v, ok := val.(float64); ok {
				vec[i] = float32(v)
			}
		}

		if len(vec) != col.Dimension() {
			continue
		}

		// 构造 Point
		point := vectordb.Point{
			ID:     id,
			Vector: vec,
		}

		// 转换 meta
		if metaRaw, ok := p["meta"]; ok && metaRaw != nil {
			if metaBytes, err := json.Marshal(metaRaw); err == nil {
				point.Meta = json.RawMessage(metaBytes)
			}
		}

		col.InsertPoint(point)
		addedCount++
	}

	// 持久化
	vectordb.SaveCollection(col, vectordb.GlobalDB.Path)

	ret["data"] = map[string]interface{}{
		"added_count": addedCount,
	}
}

func vectorDelete(c *gin.Context) {
	ensureVectorDB()

	ret := map[string]interface{}{"code": 0}
	defer c.JSON(http.StatusOK, ret)

	body := map[string]interface{}{}
	if err := c.BindJSON(&body); err != nil {
		ret["code"] = -1
		ret["msg"] = "参数解析失败"
		return
	}

	collectionName, _ := body["collection_name"].(string)
	idsRaw, _ := body["ids"].([]interface{})

	// 保护 embedding 专用集合
	if isEmbeddingReservedCollection(collectionName) {
		ret["code"] = 403
		ret["msg"] = "Embedding 集合受保护，请使用 /api/embedding/* 接口"
		return
	}

	col := vectordb.GlobalDB.GetCollection(collectionName)
	if col == nil {
		ret["code"] = -1
		ret["msg"] = "数据集不存在"
		return
	}

	deletedCount := 0
	for _, idRaw := range idsRaw {
		if id, ok := idRaw.(string); ok {
			col.DeleteItemWithIndex(id)
			deletedCount++
		}
	}

	// 持久化
	vectordb.SaveCollection(col, vectordb.GlobalDB.Path)

	ret["data"] = map[string]interface{}{
		"deleted_count": deletedCount,
	}
}

func vectorQuery(c *gin.Context) {
	ensureVectorDB()

	ret := map[string]interface{}{"code": 0}
	defer c.JSON(http.StatusOK, ret)

	body := map[string]interface{}{}
	if err := c.BindJSON(&body); err != nil {
		ret["code"] = -1
		ret["msg"] = "参数解析失败"
		return
	}

	collectionName, _ := body["collection_name"].(string)
	vectorRaw, _ := body["vector"].([]interface{})
	topK := 10
	if k, ok := body["top_k"].(float64); ok && k > 0 {
		topK = int(k)
	}
	// 兼容旧参数名 limit
	if k, ok := body["limit"].(float64); ok && k > 0 {
		topK = int(k)
	}
	efSearch := 0
	if ef, ok := body["ef_search"].(float64); ok {
		efSearch = int(ef)
	}

	col := vectordb.GlobalDB.GetCollection(collectionName)
	if col == nil {
		ret["code"] = -1
		ret["msg"] = "数据集不存在"
		return
	}

	// 转换向量
	queryVec := make([]float32, len(vectorRaw))
	for i, val := range vectorRaw {
		if v, ok := val.(float64); ok {
			queryVec[i] = float32(v)
		}
	}

	if len(queryVec) != col.Dimension() {
		ret["code"] = -1
		ret["msg"] = "向量维度不匹配"
		return
	}

	results := col.Search(queryVec, topK, efSearch)
	ret["data"] = results
}

func vectorKeys(c *gin.Context) {
	ensureVectorDB()

	ret := map[string]interface{}{"code": 0}
	defer c.JSON(http.StatusOK, ret)

	body := map[string]interface{}{}
	if err := c.BindJSON(&body); err != nil {
		ret["code"] = -1
		ret["msg"] = "参数解析失败"
		return
	}

	collectionName, _ := body["collection_name"].(string)
	withMeta, _ := body["with_meta"].(bool)

	col := vectordb.GlobalDB.GetCollection(collectionName)
	if col == nil {
		ret["code"] = -1
		ret["msg"] = "数据集不存在"
		return
	}

	if withMeta {
		var result []map[string]interface{}
		col.ForEachID(func(id string, _ uint64, metaBytes []byte) bool {
			item := map[string]interface{}{"id": id}
			if len(metaBytes) > 0 {
				var metaObj interface{}
				if json.Unmarshal(metaBytes, &metaObj) == nil {
					item["meta"] = metaObj
				}
			}
			result = append(result, item)
			return true
		})
		ret["data"] = result
	} else {
		keys := col.ListIDs()
		ret["data"] = keys
	}
}

func vectorState(c *gin.Context) {
	ensureVectorDB()

	ret := map[string]interface{}{"code": 0}
	defer c.JSON(http.StatusOK, ret)

	body := map[string]interface{}{}
	if err := c.BindJSON(&body); err != nil {
		ret["code"] = -1
		ret["msg"] = "参数解析失败"
		return
	}

	collectionName, _ := body["collection_name"].(string)

	col := vectordb.GlobalDB.GetCollection(collectionName)
	if col == nil {
		ret["code"] = -1
		ret["msg"] = "数据集不存在"
		return
	}

	ret["data"] = map[string]interface{}{
		"name":       col.Info().Name,
		"dimension":  col.Info().Dimension,
		"item_count": col.ItemCount(),
	}
}

func vectorRebuild(c *gin.Context) {
	ensureVectorDB()

	ret := map[string]interface{}{"code": 0}
	defer c.JSON(http.StatusOK, ret)

	body := map[string]interface{}{}
	if err := c.BindJSON(&body); err != nil {
		ret["code"] = -1
		ret["msg"] = "参数解析失败"
		return
	}

	collectionName, _ := body["collection_name"].(string)

	col := vectordb.GlobalDB.GetCollection(collectionName)
	if col == nil {
		ret["code"] = -1
		ret["msg"] = "数据集不存在"
		return
	}

	if err := col.RebuildIndex(); err != nil {
		ret["code"] = -1
		ret["msg"] = err.Error()
		return
	}

	// 持久化
	vectordb.SaveCollection(col, vectordb.GlobalDB.Path)

	ret["msg"] = "索引重建完成"
}

func vectorDeleteCollection(c *gin.Context) {
	ensureVectorDB()

	ret := map[string]interface{}{"code": 0}
	defer c.JSON(http.StatusOK, ret)

	body := map[string]interface{}{}
	if err := c.BindJSON(&body); err != nil {
		ret["code"] = -1
		ret["msg"] = "参数解析失败"
		return
	}

	collectionName, _ := body["collection_name"].(string)

	if collectionName == "" {
		ret["code"] = -1
		ret["msg"] = "collection_name 不能为空"
		return
	}

	// 保护 embedding 专用集合
	if isEmbeddingReservedCollection(collectionName) {
		ret["code"] = 403
		ret["msg"] = "Embedding 集合受保护，请使用 /api/embedding/* 接口"
		return
	}

	col := vectordb.GlobalDB.GetCollection(collectionName)
	if col == nil {
		ret["code"] = -1
		ret["msg"] = "数据集不存在"
		return
	}

	// 删除集合
	if err := vectordb.GlobalDB.DeleteCollection(collectionName); err != nil {
		ret["code"] = -1
		ret["msg"] = "删除数据集失败: " + err.Error()
		return
	}

	ret["msg"] = "数据集已删除"
	ret["data"] = map[string]interface{}{
		"collection_name": collectionName,
	}
}

File: kernelSDKTS/src/apiDefs/vector.ts
/**
 * 向量数据库相关 API 定义
 * 
 * 这些 API 用于操作通用向量集合。
 * 注意：blocks_embedding_* 和 assets_embedding_* 前缀的集合是 embedding 专用集合，
 * 只能通过 /api/embedding/* 接口操作，不能通过这些通用 vector API 直接操作。
 */
import { z } from 'zod';
import type { Api定义 } from '../client/types';
import { 创建响应Schema } from './types';

/**
 * 向量点 Schema
 */
const VectorPointSchema = z.object({
    id: z.string().describe('向量点的唯一标识符'),
    vector: z.array(z.number()).describe('向量数据，float32 数组'),
    meta: z.any().optional().describe('可选的元数据，任意 JSON 对象'),
});

/**
 * 搜索结果 Schema
 */
const SearchResultSchema = z.object({
    id: z.string().describe('匹配的向量点 ID'),
    score: z.number().describe('匹配得分/相似度'),
    meta: z.any().optional().describe('关联的元数据'),
});

export const vectorApiDefs = [
    {
        method: 'POST',
        endpoint: '/api/vector/collections/build',
        en: 'vectorBuildCollection',
        zh_cn: '创建向量集合',
        description: '创建一个新的向量集合。集合名称不能以 blocks_embedding_ 或 assets_embedding_ 开头，这些是 embedding 专用集合。',
        needAuth: true,
        needAdminRole: true,
        unavailableIfReadonly: true,
        zodRequestSchema: z.object({
            collection_name: z.string().describe('集合名称，不能为空'),
            dimension: z.number().int().positive().describe('向量维度，必须大于 0'),
        }),
        zodResponseSchema: 创建响应Schema(
            z.object({
                collection_name: z.string().describe('创建的集合名称'),
                dimension: z.number().describe('向量维度'),
            }).nullable()
        ),
        lastVerified: '2025-12-29',
    },
    {
        method: 'POST',
        endpoint: '/api/vector/collections/delete',
        en: 'vectorDeleteCollection',
        zh_cn: '删除向量集合',
        description: '删除指定的向量集合及其所有数据。集合名称不能以 blocks_embedding_ 或 assets_embedding_ 开头。',
        needAuth: true,
        needAdminRole: true,
        unavailableIfReadonly: true,
        zodRequestSchema: z.object({
            collection_name: z.string().describe('要删除的集合名称'),
        }),
        zodResponseSchema: 创建响应Schema(
            z.object({
                collection_name: z.string().describe('已删除的集合名称'),
            }).nullable()
        ),
        lastVerified: '2025-12-29',
    },
    {
        method: 'POST',
        endpoint: '/api/vector/add',
        en: 'vectorAdd',
        zh_cn: '添加向量点',
        description: '向集合中添加一个或多个向量点。每个向量点需包含 id、vector 和可选的 meta。',
        needAuth: true,
        needAdminRole: true,
        unavailableIfReadonly: true,
        zodRequestSchema: z.object({
            collection_name: z.string().describe('目标集合名称'),
            points: z.array(VectorPointSchema).describe('要添加的向量点数组'),
        }),
        zodResponseSchema: 创建响应Schema(
            z.object({
                added_count: z.number().describe('成功添加的向量点数量'),
            }).nullable()
        ),
        lastVerified: '2025-12-29',
    },
    {
        method: 'POST',
        endpoint: '/api/vector/delete',
        en: 'vectorDelete',
        zh_cn: '删除向量点',
        description: '根据 ID 列表删除集合中的向量点。',
        needAuth: true,
        needAdminRole: true,
        unavailableIfReadonly: true,
        zodRequestSchema: z.object({
            collection_name: z.string().describe('目标集合名称'),
            ids: z.array(z.string()).describe('要删除的向量点 ID 列表'),
        }),
        zodResponseSchema: 创建响应Schema(
            z.object({
                deleted_count: z.number().describe('成功删除的向量点数量'),
            }).nullable()
        ),
        lastVerified: '2025-12-29',
    },
    {
        method: 'POST',
        endpoint: '/api/vector/query',
        en: 'vectorQuery',
        zh_cn: '向量相似性搜索',
        description: '在集合中搜索与给定向量最相似的 top_k 个结果。',
        needAuth: true,
        needAdminRole: false,
        unavailableIfReadonly: false,
        zodRequestSchema: z.object({
            collection_name: z.string().describe('目标集合名称'),
            vector: z.array(z.number()).describe('查询向量'),
            top_k: z.number().int().positive().optional().describe('返回最相似的前 k 个结果，默认为 10'),
            limit: z.number().int().positive().optional().describe('top_k 的别名，兼容旧参数'),
            ef_search: z.number().int().optional().describe('HNSW 搜索参数，越大越精确但越慢'),
        }),
        zodResponseSchema: 创建响应Schema(
            z.array(SearchResultSchema).describe('搜索结果数组')
        ),
        lastVerified: '2025-12-29',
    },
    {
        method: 'POST',
        endpoint: '/api/vector/keys',
        en: 'vectorKeys',
        zh_cn: '获取向量点列表',
        description: '获取集合中所有向量点的 ID 列表，可选择是否包含元数据。',
        needAuth: true,
        needAdminRole: false,
        unavailableIfReadonly: false,
        zodRequestSchema: z.object({
            collection_name: z.string().describe('目标集合名称'),
            with_meta: z.boolean().optional().describe('是否同时返回元数据，默认为 false'),
        }),
        zodResponseSchema: 创建响应Schema(
            z.union([
                z.array(z.string()).describe('仅 ID 列表 (when with_meta=false)'),
                z.array(z.object({
                    id: z.string().describe('向量点 ID'),
                    meta: z.any().optional().describe('元数据'),
                })).describe('包含元数据的对象数组 (when with_meta=true)'),
            ])
        ),
        lastVerified: '2025-12-29',
    },
    {
        method: 'POST',
        endpoint: '/api/vector/state',
        en: 'vectorState',
        zh_cn: '获取集合状态',
        description: '获取向量集合的当前状态信息，包括维度、记录数等。',
        needAuth: true,
        needAdminRole: false,
        unavailableIfReadonly: false,
        zodRequestSchema: z.object({
            collection_name: z.string().describe('目标集合名称'),
        }),
        zodResponseSchema: 创建响应Schema(
            z.object({
                name: z.string().describe('集合名称'),
                dimension: z.number().describe('向量维度'),
                item_count: z.number().describe('向量点数量'),
                max_layer: z.number().describe('HNSW 索引最大层数'),
            }).nullable()
        ),
        lastVerified: '2025-12-29',
    },
    {
        method: 'POST',
        endpoint: '/api/vector/rebuild',
        en: 'vectorRebuild',
        zh_cn: '重建索引',
        description: '重建集合的 HNSW 索引。在大量增删操作后可以重建索引来优化搜索性能。',
        needAuth: true,
        needAdminRole: true,
        unavailableIfReadonly: true,
        zodRequestSchema: z.object({
            collection_name: z.string().describe('目标集合名称'),
        }),
        zodResponseSchema: 创建响应Schema(z.null()),
        lastVerified: '2025-12-29',
    },
] as const satisfies readonly Api定义[];

export type VectorApiDefs = typeof vectorApiDefs;
